src/mpid/ch4/netmod/ofi/ofi_events.c

```diff
@@ -322,25 +349,49 @@ int MPIDI_OFI_get_huge_event(struct fi_cq_tagged_entry *wc, MPIR_Request * req)
             goto fn_exit;
         }
 
-        remote_key = recv_elem->remote_info.rma_key;
+        int nic = 0;
         int vni_src = recv_elem->remote_info.vni_src;
         int vni_dst = recv_elem->remote_info.vni_dst;
-        int vni_local = vni_dst;
-        int vni_remote = vni_src;
-        int nic = 0;
-        int ctx_idx = MPIDI_OFI_get_ctx_index(vni_local, nic);
-
-        MPIDI_OFI_cntr_incr(vni_local, nic);
-        MPIDI_OFI_CALL_RETRY(fi_read(MPIDI_OFI_global.ctx[ctx_idx].tx,  /* endpoint     */
-                                     (void *) ((uintptr_t) recv_elem->wc.buf + recv_elem->cur_offset),  /* local buffer */
-                                     bytesToGet,        /* bytes        */
-                                     NULL,      /* descriptor   */
-                                     MPIDI_OFI_comm_to_phys(recv_elem->comm_ptr, recv_elem->remote_info.origin_rank, nic, vni_local, vni_remote),       /* Destination  */
-                                     recv_rbase(recv_elem) + recv_elem->cur_offset,     /* remote maddr */
-                                     remote_key,        /* Key          */
-                                     (void *) &recv_elem->context), vni_local, rdma_readfrom,   /* Context */
-                             FALSE);
-        recv_elem->cur_offset += bytesToGet;
+        if (MPIDI_OFI_COMM(recv_elem->comm_ptr).enable_striping) {      /* if striping enabled */
+            MPIDI_OFI_cntr_incr(vni_src, nic);
+            if (recv_elem->cur_offset >= MPIDI_OFI_STRIPE_CHUNK_SIZE && bytesLeft > 0) {
+                for (nic = 0; nic < MPIDI_OFI_global.num_nics; nic++) {
+                    int ctx_idx = MPIDI_OFI_get_ctx_index(vni_dst, nic);
+                    remote_key = recv_elem->remote_info.rma_keys[nic];
```

Comment:
`nic` here is the local NIC index, right? So this code assumes the same number of NICs on each process? Or can some asymmetric configurations work? E.g. if the receiver has fewer NICs than the sender, this should be OK. But if the receiver has more, this array access may be a problem. Is there any safeguard?

Reply:
The assumption is that there are the same numbers of NICs on all of the nodes. This assumption was built into the previous PR (https://github.com/pmodels/mpich/pull/5154) as well.

Reply:
I wasn't quite aware of this assumption. Do we have an assertion to make this assumption explicit?

Reply:
I don't think we have an assertion for that anywhere. It would probably need to happen during `MPIDI_OFI_post_init` before the call to `addr_exchange_all_vnis` (in case one node has one NIC and others have more than one).

Reply:
Yes, I think we need to do that. We probably also need to check the subnets since apparently that is an issue.

Reply:
I don't know how to reliably check that the NICs can communicate. I'm fine with adding the check for the same number of NICs, but any subnet check I can think of would be extremely hacky (especially inside MPICH).

Reply:
I guess I am ok with a `FIXME` comment for now if a good solution is not immediately clear.

Reply:
I've added a check for that now.